src/inmanta/data/__init__.py

```diff
@@ -4840,6 +4840,387 @@ def get_compliance_status(self) -> Optional[state.Compliance]:
+                    $1,
+                    $2,
+                    irs.id
+                FROM inserted_resource_sets AS irs
+            ), resource_data AS (
+                SELECT
+                    UNNEST($4::text[]) AS resource_id,
+                    UNNEST($5::text[]) AS resource_type,
+                    UNNEST($6::text[]) AS resource_id_value,
+                    UNNEST($7::text[]) AS agent,
+                    UNNEST($8::jsonb[]) AS attributes,
+                    UNNEST($9::text[]) AS attribute_hash,
+                    UNNEST($10::boolean[]) AS is_undefined,
+                    UNNEST($11::text[]) AS resource_set
+            )
+            -- insert resources
+            INSERT INTO public.resource(
+                environment,
+                model,
+                resource_id,
+                resource_type,
+                resource_id_value,
+                agent,
+                attributes,
+                attribute_hash,
+                is_undefined,
+                resource_set,
+                resource_set_id
+            )
+            SELECT
+                $1,
+                $2,
+                r.resource_id,
+                r.resource_type,
+                r.resource_id_value,
+                r.agent,
+                r.attributes,
+                r.attribute_hash,
+                r.is_undefined,
+                rs.name,
+                rs.id
+            FROM resource_data AS r
+            -- this join has been tested to be up to four times faster than joining with
+            -- resource_configuration_model, even if the latter would have the name column directly
+            -- (for 5k models, 5k sets, updating 1-1000 sets, with 100-10k resources per set).
+            -- Order of magnitude for reference: 0.5s when updating 10 sets with 1k resources per set.
+            INNER JOIN inserted_resource_sets AS rs
+                ON r.resource_set IS NOT DISTINCT FROM rs.name
+            """,
+            *common_values,
+            cls._get_value(updated_resource_sets),
+            # insert as separate arrays rather than jsonb because jsob has a size limit
+            resource_data_db.get("resource_id", []),
+            resource_data_db.get("resource_type", []),
+            resource_data_db.get("resource_id_value", []),
+            resource_data_db.get("agent", []),
+            resource_data_db.get("attributes", []),
+            resource_data_db.get("attribute_hash", []),
+            resource_data_db.get("is_undefined", []),
+            resource_data_db.get("resource_set", []),
+            connection=connection,
+        )
+
+        if is_partial_update:
+            await cls.validate_resource_sets_in_version(
+                environment=environment,
+                version=target_version,
+                updated_resource_sets=updated_resource_sets,
+                connection=connection,
+            )
+
+    @classmethod
+    async def clear_resource_sets(
```

Comment:
Logically this query looks good to me. Something I noticed is that the locking order between the case where the version argument is set or not set is not the same, because the case with the version argument does an additional select query:

Locking order with version argument:

* resource_set_configuration_model
* resource_set
* resource

Without version argument it's:
* resource_set_configuration_model
* resource
* resource_set

We had issues with locking order in the past, so we try to keep it stable. But maybe this will not cause any issues in practice. It's hard to estimate.